insolver/feature_monitoring/homogeneity_report.py

```diff
@@ -0,0 +1,313 @@
+
+        # check template file
+        if render:
+            if report_path is None:
+                report_path = 'homogeneity_report.html'
+            dir = ntpath.dirname(inspect.getfile(HomogeneityReport))
+            template_path = dir + '/' + 'report_template.html'
+            if not os.path.exists(template_path):
+                raise OSError("Can not find template file. It must be in 'feature_monitoring' package.")
+
+        # raise warning if function gets output path but rendering is not asked
+        if not render and not (report_path is None):
+            raise Warning("Argument 'report_path' will be ignored as report rendering is not asked.")
+
+        # accurately assemble report data
+        report_data = []
+        for feat in features:
+            properties = self.config_dict[feat]
+
+            # required parameter
+            feat_type = properties['feature_type']
+
+            # check optional stat. parameters
+            pval_thresh = 0.05 if ('pval_thresh' not in properties) else properties['pval_thresh']
+            samp_size = 500 if ('samp_size' not in properties) else properties['samp_size']
+            bootstrap_num = 100 if ('bootstrap_num' not in properties) else properties['bootstrap_num']
+
+            # count nan difference between x1, x2
+            nan_perc1 = df1[feat].isna().sum() / len(df1[feat])
+            nan_perc2 = df2[feat].isna().sum() / len(df2[feat])
+            nan_perc_gap = nan_perc2 - nan_perc1
+            nan_gap_dict = {'nan_perc1': nan_perc1, 'nan_perc2': nan_perc2, 'nan_perc_gap': nan_perc_gap}
+
+            # copy data to avoid side effects
+            if dropna:
+                x1 = df1[feat].dropna().values
+                x2 = df2[feat].dropna().values
+            else:
+                x1 = df1[feat].values.copy()
+                x2 = df2[feat].values.copy()
+
+            if feat_type == 'continuous':
+                # optional psi_bins
+                psi_bins = 20 if ('psi_bins' not in properties) else properties['psi_bins']
+
+                # if we need to render report we build charts
+                if render:
+                    x1, x2, _ = fillna_cont(x1, x2, inplace=True)
+                    chart_bins = 15 if ('chart_bins' not in properties) else properties['chart_bins']
+                    if not ('chart_limits' in properties):
+                        chart_limits = min(x1.min(), x2.min()), max(x2.max(), x2.max())
+                    else:
+                        chart_limits = properties['chart_limits']
+
+                    chart = chart_cont(x1, x2, name1, name2, chart_limits, chart_bins, offline=True)
+
+                # run tests
+                homogen_tester = ContinuousHomogeneityTests(pval_thresh, samp_size, bootstrap_num, psi_bins)
+                test_results = homogen_tester.run_all(x1, x2, inplace=True)
+
+            elif feat_type == 'discrete':
```

Comment:
It is better to specify the error when `feat_type` is neither `continuous` nor `discrete`

Reply:
Good note. This property sits inside config. I decided to make config_dict private (--> __config_dict). This check moved to setter and also __init__ turned to setter calling. So we always check this error on config modifying (and some others).